x/logic/predicate/file_test.go

```diff
@@ -0,0 +1,128 @@
+//nolint:gocognit
+package predicate
+
+import (
+	"fmt"
+	"net/url"
+	"testing"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/store"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/golang/mock/gomock"
+	"github.com/ichiban/prolog/engine"
+	"github.com/okp4/okp4d/x/logic/fs"
+	"github.com/okp4/okp4d/x/logic/testutil"
+	"github.com/okp4/okp4d/x/logic/types"
+	. "github.com/smartystreets/goconvey/convey"
+	"github.com/tendermint/tendermint/libs/log"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+	tmdb "github.com/tendermint/tm-db"
+)
+
+func TestSourceFile(t *testing.T) {
+	Convey("Given test cases", t, func() {
+		ctrl := gomock.NewController(t)
+		defer ctrl.Finish()
+
+		cases := []struct {
+			query       string
+			wantResult  []types.TermResults
+			wantError   error
+			wantSuccess bool
+		}{
+			{
+				query:       "source_file(file).",
+				wantSuccess: false,
+			},
+			{
+				query:       "consult(file1), consult(file2), source_file(file1).",
+				wantResult:  []types.TermResults{{}},
+				wantSuccess: true,
+			},
+			{
+				query:       "consult(file1), consult(file2), source_file(file3).",
+				wantSuccess: false,
+			},
+			{
+				query:       "source_file(X).",
+				wantSuccess: false,
+			},
+			{
+				query:       "consult(file1), consult(file2), source_file(X).",
+				wantResult:  []types.TermResults{{"X": "file1"}, {"X": "file2"}},
+				wantSuccess: true,
+			},
```

Comment:
If the returned sources are sorted, add a test case which shows it. For example:
```go
          {
				query:       "consult(file2), consult(file3), consult(file1), source_file(X).",
				wantResult:  []types.TermResults{{"X": "file1"}, {"X": "file2"}, {"X": "file3"}},
				wantSuccess: true,
			},
```